### PR TITLE
Restore support for --cache command line flag

### PR DESCRIPTION
**Pull request by juj**
Restore support for --cache command line flag to configure the location of the Emscripten cache root directory.

Some build systems may not be well designed around the concept that one needs to configure environment variables to drive the compiler, but passing command line arguments is natural for them instead.

Also fix too late importing of toolchain profiler, which would cause it to miss time, and fix --em-config and --cache from response files.

### CODE REVIEW

**Comment by sbc100** [comment]
Sorry I broke your setup!   Can I ask what you use case is for wanting to specify this on a per-command basis rather than via the config file?

File: emcc.py [hunk 1]
```diff
@@ -37,24 +41,28 @@
...
 from subprocess import PIPE
 
+from tools.response_file import substitute_response_files
+
+# read response files very early on so that all subsequent code that accesses
+# sys.argv will see the expanded content.
+try:
+  sys.argv = substitute_response_files(sys.argv)
+except IOError as e:
+  raise Exception('Unable to parse response files!\n' + str(e))
```
Comment (sbc100) [inline emcc.py:51]: I wonder if its worth moving this into `config.py` so it applies equally to all commands?

Comment (sbc100) [reply emcc.py:51]: And presumably if we do this here we should avoid doing it again down in the `run` function?

**Comment by juj** [comment]
> Sorry I broke your setup! Can I ask what you use case is for wanting to specify this on a per-command basis rather than via the config file?

There are a relatively large number of developers who have both Unity and Emscripten SDK installed. When they use Unity to build to WebGL, they do not expect that just having Emscripten SDK installed and its env. vars activated at the system level would break Unity from working (and using its bundled Emscripten).

We have been using the command line variables `--em-config` and `--cache` to specify the location of the needed configuration items to drive the Emscripten compiler invocations. We do specify the CACHE location in .emscripten config variable, but cannot rely on it, because if we did, developers having a custom `EM_CACHE` env. var set would override the .emscripten config location, breaking Unity WebGL compiles.

This has also kept the build system at parity to other platforms which do not typically require configuring environment variables in order to be able to drive the compiler.

File: emcc.py (hunk 1 above)
Comment (juj) [reply emcc.py]: > I wonder if its worth moving this into `config.py` so it applies equally to all commands?

Works for me. Moved it over.

> And presumably if we do this here we should avoid doing it again down in the `run` function?

Good catch, forgot to remove that. Updated.

Comment (juj) [reply emcc.py]: > > I wonder if its worth moving this into `config.py` so it applies equally to all commands?
> 
> Works for me. Moved it over.

This did break emar.py response file handling, now needs #13982 to lands first.

**Comment by sbc100** [comment]
> > Sorry I broke your setup! Can I ask what you use case is for wanting to specify this on a per-command basis rather than via the config file?
> 
> There are a relatively large number of developers who have both Unity and Emscripten SDK installed. When they use Unity to build to WebGL, they do not expect that just having Emscripten SDK installed and its env. vars activated at the system level would break Unity from working (and using its bundled Emscripten).
> 
> We have been using the command line variables `--em-config` and `--cache` to specify the location of the needed configuration items to drive the Emscripten compiler invocations. We do specify the CACHE location in .emscripten config variable, but cannot rely on it, because if we did, developers having a custom `EM_CACHE` env. var set would override the .emscripten config location, breaking Unity WebGL compiles.
> 
> This has also kept the build system at parity to other platforms which do not typically require configuring environment variables in order to be able to drive the compiler.

I see.   To me it sounds like that problem stems from the fact that emsdk sets `EM_CACHE` in the environment.    I suggest one of two fixes:

a) Stop setting `EM_CACHE` as part of emsdk.   This is redundant, and we should not need to do it since the cache is configured via the config file.  I think there may be historical reasons why do this but I can't remember them right now.   We should revisit.

b) Have a say to signal to emscripten that the config file is final.   Perhaps `--hermeric` or `--ignore-env`.   We have had a request like this from our internal users who doesn't like the way the user's environment can effect builds.   Perhaps the use of `--em-config` itself should be enough to signal this?    See https://github.com/emscripten-core/emscripten/issues/12611.   This sounds very similar to what your are asking for.

**Comment by juj** [comment]
> I see. To me it sounds like that problem stems from the fact that emsdk sets `EM_CACHE` in the environment. I suggest one of two fixes:
> 
> a) Stop setting `EM_CACHE` as part of emsdk. This is redundant, and we should not need to do it since the cache is configured via the config file. I think there may be historical reasons why do this but I can't remember them right now. We should revisit.
> 
> b) Have a say to signal to emscripten that the config file is final. Perhaps `--hermeric` or `--ignore-env`. We have had a request like this from our internal users who doesn't like the way the user's environment can effect builds. Perhaps the use of `--em-config` itself should be enough to signal this? See #12611. This sounds very similar to what your are asking for.

I was pondering about a) also, but I don't think that is a good solution, since it would be removing another feature even. Also b) seems awkward as well. To me the simplest way will be to just keep supporting `--cache` cmdline var, it is very symmetric to `--em-config` cmdline var, there are the three locations to configure: cmdline, env.var, config file (in that order).

(also earlier you had landed the patch that all .emscripten config file variables can symmetrically be defined as env. vars. Removing that support specifically for `EM_CACHE` would be a bit arbitrary compared to all the others).

What was the rationale for needing to remove `--cache` variable? It worked well, was useful, and I don't think it was a maintenance burden.

**Comment by sbc100** [comment]
> > I see. To me it sounds like that problem stems from the fact that emsdk sets `EM_CACHE` in the environment. I suggest one of two fixes:
> > a) Stop setting `EM_CACHE` as part of emsdk. This is redundant, and we should not need to do it since the cache is configured via the config file. I think there may be historical reasons why do this but I can't remember them right now. We should revisit.
> > b) Have a say to signal to emscripten that the config file is final. Perhaps `--hermeric` or `--ignore-env`. We have had a request like this from our internal users who doesn't like the way the user's environment can effect builds. Perhaps the use of `--em-config` itself should be enough to signal this? See #12611. This sounds very similar to what your are asking for.
> 
> I was pondering about a) also, but I don't think that is a good solution, since it would be removing another feature even. Also b) seems awkward as well. To me the simplest way will be to just keep supporting `--cache` cmdline var, it is very symmetric to `--em-config` cmdline var, there are the three locations to configure: cmdline, env.var, config file (in that order).
> 
> (also earlier you had landed the patch that all .emscripten config file variables can symmetrically be defined as env. vars. Removing that support specifically for `EM_CACHE` would be a bit arbitrary compared to all the others).
> 

I'm not proposing removing support for `EM_CACHE` .. just that emsdk should not be setting it (it uses the config file to do all its configuring and we should avoid polluting the environment when a config var works just as well).   

> What was the rationale for needing to remove `--cache` variable? It worked well, was useful, and I don't think it was a maintenance burden.

It just seemed redundant, and weird, that all the other config variables are settable via the env or file, but this one was special and had it own command line too.     We don't have corresponding command line args for `--nodejs` or `--llvm-root` for example.

Also, if we can avoid you needing to add `--cache` in unity its just one less thing for you to worry about.  If my change to emsdk lands (https://github.com/emscripten-core/emsdk/pull/797) then you just won't need this extra command line at all.

**Comment by sbc100** [comment]
It seems there are two different sides to this issue:

1. User's environment variables can interfere with otherwise per-packaged SDK. (EM_LLVM, EM_NODE_JS, etc)
2. emsdk sets one of these which means that such inferences is common/surprising for normal users.

I would say that (2) seems like the real problem here.   

It might also be nice to solve (1), but its not very common I would imagine.  Is it ever useful for you as a dev who works on unity to be able to set `EM_LLVM` in order to override the version of llvm use in the pre-packaged version?  Maybe?  Maybe (1) could be a feature in some cases but (2) certainly seems like a bug.

**Comment by juj** [comment]
> I'm not proposing removing support for `EM_CACHE` .. just that emsdk should not be setting it (it uses the config file to do all its configuring and we should avoid polluting the environment when a config var works just as well).

> Also, if we can avoid you needing to add `--cache` in unity its just one less thing for you to worry about. If my change to emsdk lands ([emscripten-core/emsdk#797](https://github.com/emscripten-core/emsdk/pull/797)) then you just won't need this extra command line at all.

The PR https://github.com/emscripten-core/emsdk/pull/797 looks fine, but it will unfortunately not change anything with respect to needing this PR: people who have existing emsdk installations will still have `EM_CACHE` set (or if they have manually set that), and they will break in the absence of `--cache` since they won't automatically get that emsdk PR applied to their existing system installations.

If we wanted to fix the issue by the route of removing, we'd have to remove the support for `EM_CACHE` env. var in Emscripten, but I don't think that is a good idea.

>     1. User's environment variables can interfere with otherwise per-packaged SDK. (EM_LLVM, EM_NODE_JS, etc)

> It might also be nice to solve (1), but its not very common I would imagine.

This is an unfortunate issue that came about after #9177 was added. I am now surprised by what prompted adding that feature, the rationale was not provided there.

However that being said, I don't think these explicit env. vars will be much of an issue for us, since if a developer has manually & deliberately added this kind of an override env. var for a special purpose, then I'd expect they would be able to tear it down. 

Rather the issue is with existing Emsdk users who might use it for some other development purposes.

>     2\. emsdk sets one of these which means that such inferences is common/surprising for normal users.
> 
> 
> I would say that (2) seems like the real problem here.

(2) is a problem, but to note, it only became a problem after --cache was removed.

>  Is it ever useful for you as a dev who works on unity to be able to set `EM_LLVM` in order to override the version of llvm use in the pre-packaged version?

If someone is a very very power user, it might be useful, but I don't currently know of any user who could do so. If they are such a super power user, they can also edit the .emscripten config file that Unity uses to point to their own LLVM.

> > What was the rationale for needing to remove `--cache` variable? It worked well, was useful, and I don't think it was a maintenance burden.
> 
> It just seemed redundant, and weird, that all the other config variables are settable via the env or file, but this one was special and had it own command line too. We don't have corresponding command line args for `--nodejs` or `--llvm-root` for example.

Even if something is a general pattern of N over a single instance of one, it might not validate the feature's existence any more or less. Hopefully "stability for historical reasons" can be considered for features even if they might be one-off, especially _if_ the maintenance burden is not particularly high. Removing `--cache` on the basis of it not being symmetric was awkwardly unpragmatic.

**Comment by sbc100** [comment]
OK you make some good arguments.   I'm OK landing this as is for now.

It agree its possible that old emsdk might have old values of `EM_CACHE` still set in their environment, and that seems like a bug in emsdk to me.   If you activate a new SDK it should not just add new values but remove any old ones from the environment.    We we can fix that (such that EM_CACHE is guaranteed not to be set for any modern emsdk activation) would you (unity) consider removing the `--cache` argument from your command line and sticking to the config files and/or defaults?

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
I opened https://github.com/emscripten-core/emsdk/issues/800 to fix the "unset old variables" issue.

**Comment by juj** [comment]
> It agree its possible that old emsdk might have old values of `EM_CACHE` still set in their environment, and that seems like a bug in emsdk to me. If you activate a new SDK it should not just add new values but remove any old ones from the environment

Originally emsdk activate was intended to add a tool, i.e. one could then type

```
emsdk activate emscripten-main-64bit
emsdk activate ninja-git-64bit
emsdk activate ...
```
to one by one activate multiple tools. If activate is changed to "set only these tools active and nothing else", then the above will no longer work. I do find that usage useful at times, but if we do want to change the semantics, that is ok too. Maybe multiple activation commands are not that commonly done anyways.

> we can fix that (such that EM_CACHE is guaranteed not to be set for any modern emsdk activation) would you (unity) consider removing the `--cache` argument from your command line and sticking to the config files and/or defaults?

We need to have a suitable amount of time pass between the removal of EM_CACHE from emsdk and the removal of --cache from Emscripten so that developers will not break easily.

**Comment by juj** [comment]
> or defaults?

The reason we do not use the default cache location is that we received a surprisingly large number of bug reports from developers who say that WebGL builds take ages to complete. It can take 10-20minutes to build the cache on weaker systems, and on top of that there will be the project build time that adds up.

Currently Tiny Unity does use the default cache location, and several developers have commented about this issue. It is not easy for them to know that first time build would spend a long time doing cache population.

**Comment by sbc100** [comment]
> > or defaults?
> 
> The reason we do not use the default cache location is that we received a surprisingly large number of bug reports from developers who say that WebGL builds take ages to complete. It can take 10-20minutes to build the cache on weaker systems, and on top of that there will be the project build time that adds up.
> 
> Currently Tiny Unity does use the default cache location, and several developers have commented about this issue. It is not easy for them to know that first time build would spend a long time doing cache population.

How is the cache location related to the whether users need to populate the cache?

I assume you are talking about shipping a pre-populated cache (as emsdk does)?  If that is the case then it seems to me that using the default location (which is inside the emscripten/cache directory) seems like a good idea, no?    I must be misunderstanding something.

**Comment by sbc100** [comment]
This seems to have broken out windows builder:

https://chromium-review.googlesource.com/c/emscripten-releases/+/2853491

```
C:/b/s/w/ir/k/install\bin\llvm-ar.exe: error: file1: no such file or directory
test_emar_response_file (test_other.other) ... ERROR
```

**Comment by sbc100** [comment]
If I try to enable this test on window CI I get this, different, error:

```
======================================================================
ERROR: test_emar_response_file (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "unittest\case.py", line 59, in testPartExecutor
  File "unittest\case.py", line 593, in run
  File "unittest\case.py", line 550, in _callTestMethod
  File "C:\Users\circleci\project\tests\test_other.py", line 7474, in test_emar_response_file
    building.emar('cr', 'libfoo.a', ("file'1", "file'2", "hyv�� p�iv��", "snowman freezes covid \u2603 \U0001f9a0"))
  File "C:\Users\circleci\project\tools\building.py", line 652, in emar
    response_filename = response_file.create_response_file(filenames, TEMP_DIR)
  File "C:\Users\circleci\project\tools\response_file.py", line 44, in create_response_file
    f.write(contents)
  File "encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode character '\u2603' in position 55: character maps to <undefined>
```

**Comment by sbc100** [comment]
My proposal for how to deal with breakage here:

1. land https://github.com/emscripten-core/emscripten/pull/14036
2. revert this change along with a re-enabling of test_emar_response_file for windows
3. re-land this change but much close to pure revert of #11094 (which we know worked)
4. Re-evaluate, as a separate change, the additional functionality that change change added which was to parse all incoming response files.   This could prove tricky.. and will probably involve writing a custom parser rather than relying on shlex.split like we do today.   In the mean time #11094 will still have been effectively reverted which was the main goal of this PR.

**Comment by sbc100** [comment]
I believe that if we land #14036 and #14039 that will take care of (1), (2) and (3) and we ca work on solving (4) with the tree green again.

**Comment by juj** [comment]
> I assume you are talking about shipping a pre-populated cache (as emsdk does)? If that is the case then it seems to me that using the default location (which is inside the emscripten/cache directory) seems like a good idea, no? I must be misunderstanding something.

We do use the cache inside the Emscripten SDK directory, but the issue is about two different installed Emsdks that people may have: their own Emsdk, and the Unity-installed Emsdk. Both may have their caches inside the Emsdk directory, but if they have their Emsdk active, then Unity Emsdk will incorrectly start using the developer's custom Emsdk directory cache instead.

> I believe that if we land #14036 and #14039 that will take care of (1), (2) and (3) and we ca work on solving (4) with the tree green again.

I am a bit pinched on time the rest of this week, but at a glance looks good. Thanks for tackling this!

**Pull request merged**